cmd/kubectl-gadget/deploy.go

```diff
@@ -362,35 +209,140 @@ func runDeploy(cmd *cobra.Command, args []string) error {
+			if err != nil {
+				return err
+			}
+			gadgetContainer.ImagePullPolicy = policy
+
+			if !livenessProbe {
+				gadgetContainer.LivenessProbe = nil
+			} else {
+				gadgetContainer.LivenessProbe.InitialDelaySeconds = livenessProbeInitialDelay
+			}
+
+			for i := range gadgetContainer.Env {
+				switch gadgetContainer.Env[i].Name {
+				case "GADGET_IMAGE":
+					gadgetContainer.Env[i].Value = image
+				case "INSPEKTOR_GADGET_VERSION":
+					gadgetContainer.Env[i].Value = version
+				case "INSPEKTOR_GADGET_OPTION_HOOK_MODE":
+					gadgetContainer.Env[i].Value = hookMode
+				case "INSPEKTOR_GADGET_OPTION_FALLBACK_POD_INFORMER":
+					gadgetContainer.Env[i].Value = strconv.FormatBool(fallbackPodInformer)
+				}
+			}
+		}
+
+		if printOnly {
+			bytes, err := yaml.Marshal(object)
+			if err != nil {
+				return fmt.Errorf("problem while marshaling object: %w", err)
+			}
+			fmt.Printf("%s---\n", bytes)
+		} else {
+			err := createResource(dynamicClient, mapper, object)
+			if err != nil {
+				return fmt.Errorf("problem while creating resource: %w", err)
+			}
+		}
+	}
+
+	if !printOnly && deployWait {
+		k8sClient, err := k8sutil.NewClientsetFromConfigFlags(utils.KubernetesConfigFlags)
+		if err != nil {
+			return utils.WrapInErrSetupK8sClient(err)
+		}
+
+		nodes, err := k8sClient.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			return fmt.Errorf("problem while listing nodes: %w", err)
+		}
+
+		// The below code (particularly how to use UntilWithSync) is highly
+		// inspired from kubectl wait source code:
+		// https://github.com/kubernetes/kubectl/blob/b5fe0f6e9c65ea95a2118746b7e04822255d76c2/pkg/cmd/wait/wait.go#L364
+		podInterface := k8sClient.CoreV1().Pods(gadgetNamespace)
+		lw := &cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.LabelSelector = "k8s-app=gadget"
+
+				return podInterface.List(context.TODO(), options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.LabelSelector = "k8s-app=gadget"
+
+				return podInterface.Watch(context.TODO(), options)
+			},
+		}
+
+		ctx, cancel := watchtools.ContextWithOptionalTimeout(context.TODO(), time.Duration(timeout)*time.Second)
+		defer cancel()
+
+		readyPods := 0
+		_, err = watchtools.UntilWithSync(ctx, lw, &v1.Pod{}, nil, func(event watch.Event) (bool, error) {
+			switch event.Type {
+			case watch.Deleted:
+				return false, fmt.Errorf("no pod in %s namespace should be deleted", gadgetNamespace)
+			case watch.Modified:
+				pod, _ := event.Object.(*v1.Pod)
+
+				if v1pod.IsPodReady(pod) {
+					readyPods++
+				}
+
+				// We cannot list the pods after we asked for their creation because the
+				// list could be empty.
+				// But, we deploy one pod per node, so we return succesfully when the
+				// number of ready pods equals the number of nodes.
+				return len(nodes.Items) == readyPods, nil
```

Comment:
I think the advantage of watching each pod is to have access to the progress. Why we are not showing it as we do for the `undeploy` command? I expect the output to be similar to the events:

```bash
$ k alpha events -n gadget pod gadget
LAST SEEN   TYPE     REASON             OBJECT             MESSAGE
5s          Normal   SuccessfulCreate   DaemonSet/gadget   Created pod: gadget-kfxwk
5s          Normal   SuccessfulCreate   DaemonSet/gadget   Created pod: gadget-pncg6
4s          Normal   Scheduled          Pod/gadget-kfxwk   Successfully assigned gadget/gadget-kfxwk to master
4s          Normal   Scheduled          Pod/gadget-pncg6   Successfully assigned gadget/gadget-pncg6 to worker
4s          Normal   Pulling            Pod/gadget-kfxwk   Pulling image "192.168.1.105:5000/gadget:francis-new-deploy"
4s          Normal   Pulled             Pod/gadget-kfxwk   Successfully pulled image "192.168.1.105:5000/gadget:francis-new-deploy" in 12.511295ms
4s          Normal   Created            Pod/gadget-kfxwk   Created container gadget
4s          Normal   Pulling            Pod/gadget-pncg6   Pulling image "192.168.1.105:5000/gadget:francis-new-deploy"
4s          Normal   Pulled             Pod/gadget-pncg6   Successfully pulled image "192.168.1.105:5000/gadget:francis-new-deploy" in 15.011706ms
4s          Normal   Created            Pod/gadget-pncg6   Created container gadget
4s          Normal   Started            Pod/gadget-pncg6   Started container gadget
4s          Normal   Started            Pod/gadget-kfxwk   Started container gadget
```

If we decide to not do that, I think it would be easier to verify the status of the daemonset instead of every single pod? I expect it to be simpler, and we could avoid relying on the number of nodes.

It would be the same as the `rollout status` command:

```bash
$ ./kubectl-gadget deploy --deploy-wait=false

$ kubectl rollout status -n gadget ds gadget --timeout=30s
Waiting for daemon set "gadget" rollout to finish: 0 of 2 updated pods are available...
Waiting for daemon set "gadget" rollout to finish: 1 of 2 updated pods are available...
daemon set "gadget" successfully rolled out
```

However, I am not sure if it verifies if the pods are ready or only if they were scheduled. 

cc @alban @mauriciovasquezbernal

Reply:
> I prefer to keep the command silent and only prints if something went wrong.
>
Actually, we agreed the opposite in our team meeting: https://github.com/kinvolk/inspektor-gadget/issues/123#issuecomment-1068950640. In addition, this is what `undeploy` is doing thus for consistency, it should do the same.

Reply:
Ok. It means we need another issue to track what this PR is not managing from what was discussed in https://github.com/kinvolk/inspektor-gadget/issues/123.